processor/src/main/resources/org/mapstruct/ap/internal/model/IterableCreation.ftl

```diff
@@ -11,6 +11,18 @@
         <@includeModel object=factoryMethod targetType=resultType/>
     <#elseif enumSet>
         EnumSet.noneOf( <@includeModel object=enumSetElementType raw=true/>.class )
+    <#elseif sourceVersionAtLeast19 && resultType.implementationType?? && ext.useSizeIfPossible?? && ext.useSizeIfPossible && canUseSize>
+        <#if linkedHashSet>
+            LinkedHashSet.newLinkedHashSet( <@iterableSize /> )
+        <#elseif linkedHashMap>
+            LinkedHashMap.newLinkedHashMap( <@iterableSize /> )
+        <#elseif hashSet>
+            HashSet.newHashSet( <@iterableSize /> )
+        <#elseif hashMap>
+            HashMap.newHashMap( <@iterableSize /> )
+        <#else>
+            new <@includeModel object=resultType.implementationType/><#if ext.useSizeIfPossible?? && ext.useSizeIfPossible && canUseSize>( <@sizeForCreation /> )<#else>()</#if>
+        </#if>
```

Comment:
I would prefer if we can simplify this. I don't like that we have 4 if checks for the different implementations. 

I think that on the `ImplementationType` we can add additional information such as the `factoryMethodName` and that is defined then we can just do `<@includeModel object=resultType.implementationType raw=true/>.${resultType.implementationType.factoryMethodName}( <@sizeForCreation /> )`

Doing it like this only the `TypeFactory` needs to be adjusted with the `ImplementationType` based on the source version. The `IterableCreation.ftl` does not need to be aware about the source version.

Reply:
@filiphr, I have refactored that code. Please, check my update.

Reply:
It seems this requested changes have been resolved